### PR TITLE
fix: should fix #714, closes out issue where length of websocket-sec-key was wrong

### LINKED ISSUE #714: Bot is unable to properly connect to certain Discord voice servers
**Git commit reference**
cd5b9cab3fb6a49ece575e0f86475268f886766b

**Describe the bug**
The problem is that `on_voice_ready` callback is not called when bot connects to certain voice channels.
Example of successful voice server connection:
1. DPP logs show this:
`DEBUG: Sending op 4 to join VC, guild xxx channel xxx`
`DEBUG: Connecting voice for guild xxx channel xxx`
`DEBUG: Connecting new voice session...`
`DEBUG: Voice websocket established; UDP endpoint: xxx.xxx.xxx.xxx:xxx [ssrc=xxx] with 7 modes`
`DEBUG: External IP address: xxx.xxx.xxx.xxx`
2. Bot connects with following request:
`GET /?v=4 HTTP/1.1`
`Host: russia8508.discord.media`
`pragma: no-cache`
`User-Agent: DPP/0.1`
`Upgrade: WebSocket`
`Connection: Upgrade`
`Sec-WebSocket-Key: xxx`
`Sec-WebSocket-Version: 13`
3. Server responds with HTTP 101 response:
`HTTP/1.1 101 Switching Protocols`
`Connection: upgrade`
`sec-websocket-accept: xxx`
`upgrade: websocket`
`CF-Cache-Status: DYNAMIC`
`Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=xxx"}],"group":"cf-nel","max_age":604800}`
`NEL: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}`
`Strict-Transport-Security: max-age=15552000; includeSubDomains; preload`
`X-Content-Type-Options: nosniff`
`Server: cloudflare`
`CF-RAY: xxx-DME`
4. Server sends JSON payload:
`1{"op":8,"d":{"v":4,"heartbeat_interval":13750.0}}`
`Connection: upgrade`
`sec-websocket-accept: xxx`
`upgrade: websocket`
`CF-Cache-Status: DYNAMIC`
`Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=xxx"}],"group":"cf-nel","max_age":604800}`
`NEL: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}`
`Strict-Transport-Security: max-age=15552000; includeSubDomains; preload`
`X-Content-Type-Options: nosniff`
`Server: cloudflare`
`CF-RAY: xxx-DME`
5. Connection proceeds. `on_voice_ready` callback is called.

Example of erroneous voice server connection in different Discord guild:
1. DPP logs show this:
`DEBUG: Sending op 4 to join VC, guild xxx channel  xxx`
`DEBUG: Connecting voice for guild xxx channel xxx`
Nothing happens after that.
2. Bot connects with following request:
`GET /?v=4 HTTP/1.1`
`Host: tel-aviv10001.discord.media`
`pragma: no-cache`
`User-Agent: DPP/0.1`
`Upgrade: WebSocket`
`Connection: Upgrade`
`Sec-WebSocket-Key: xxx`
`Sec-WebSocket-Version: 13`
3. Server responds with HTTP 400 response:
`HTTP/1.1 400 Bad request`
`Content-Type: text/html`
`Transfer-Encoding: chunked`
`Connection: keep-alive`
`Cache-Control: no-cache`
`CF-Cache-Status: DYNAMIC`
`Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=xxx"}],"group":"cf-nel","max_age":604800}`
`NEL: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}`
`Strict-Transport-Security: max-age=15552000; includeSubDomains; preload`
`X-Content-Type-Options: nosniff`
`Server: cloudflare`
`CF-RAY: xxx-DME`
`5a`
`<html><body><h1>400 Bad request</h1>`
`Your browser sent an invalid request.`
`</body></html>`
4. By design of `dpp::discord_voice_client::thread_run()` function bot tries to reconnect. Steps 2 and 3 are repeated infinitely.
It may be important to mention that in Discord client bot actually connects to voice channel and sits there even when HTTP 400 error occurs.

**To Reproduce**
Issue appears to only occur when bot connects to `tel-avivXXXXX.discord.media` Discord voice servers. The problematic servers do work in Discord desktop clients, so I think the problem is with the library.
1. Connect to voice channel that is served by `tel-avivXXXXX.discord.media` voice server.
2. Make the bot connect to it, for example call `dpp::guild::connect_member_voice()`.
3. Observe that `on_voice_ready` callback is never called although bot connects to voice channel in Discord UI.

**Expected behavior**
Bot should properly connect to voice server and `on_voice_ready` callback should be called.

**System Details:**
 - Tried it on both 
`Linux PC 5.15.90.1-microsoft-standard-WSL2 #1 SMP x86_64 x86_64 x86_64 GNU/Linux`
and
`Linux rpi4 6.1.21-v8+ #1642 SMP PREEMPT aarch64 GNU/Linux`
OSes.
 - Desktop Discord Client was used for testing

Some of my assumptions/terms may be wrong because I tried to get how DPP source code works manually.

### PR DESCRIPTION
**Pull request by braindigitalis**
## Code change checklist

- [x] I have ensured that all methods and functions are **fully documented** using doxygen style comments.
- [x] My code follows the [coding style guide](https://dpp.dev/coding-standards.html).
- [x] I tested that my change works before raising the PR.
- [x] I have ensured that I did not break any existing API calls.
- [x] I have not built my pull request using AI, a static analysis tool or similar without any human oversight.

### CODE REVIEW

**Review by Jaskowicz1 — COMMENTED** [review]
Changes look alright, will do tests to see if it works.

**Review by Neko-Life — COMMENTED** [review]
LGTM

File: src/dpp/wsclient.cpp [hunk 1]
```diff
@@ -43,7 +43,17 @@ websocket_client::websocket_client(const std::string &hostname, const std::strin
 	path(urlpath),
 	data_opcode(opcode)
 {
-	key = std::to_string(time(nullptr));
+	uint64_t k = (time(nullptr) * time(nullptr));
+	/* A 64 bit value as hex with leading zeroes is always 16 chars.
+	 *
+	 * The request MUST include a header field with the name
+	 * |Sec-WebSocket-Key|.  The value of this header field MUST be a
+	 * nonce consisting of a randomly selected 16-byte value that has
```
Comment (Mishura4) [inline src/dpp/wsclient.cpp:51]: But this doesn't make it 16-byte?

File: src/dpp/wsclient.cpp [hunk 2]
```diff
@@ -43,7 +43,17 @@ websocket_client::websocket_client(const std::string &hostname, const std::strin
...
+	 * been base64-encoded (see [Section 4 of
+	 * [RFC4648]](https://datatracker.ietf.org/doc/html/rfc4648#section-4)).
+	 * The nonce MUST be selected randomly for each connection.
+	 */
+	key = to_hex<uint64_t>(k);
```
Comment (Mishura4) [inline src/dpp/wsclient.cpp:56]: Here you should check the length and fill with 0s if it's less than 16. There is no guarantee it will be 16 otherwise

**Review by Mishura4 — CHANGES_REQUESTED** [review]

**Review by Mishura4 — APPROVED** [review]

File: src/dpp/wsclient.cpp (hunk 1 above)
Comment (braindigitalis) [reply src/dpp/wsclient.cpp:51]: as discussed to_hex will make it 16 characters by adding leading zeros up to 2x sizeof T, so for uint64_t, 16 chars

**Comment by Jaskowicz1** [comment]
Discord API keeps going down, meaning a test can't be properly conducted. I've done as much as was possible.

```
Region          | Server            | Response
---------------------------------------------------------------------
Brazil          | brazil7074        | 101 Switching Protocols (271ms)
Hong Kong       | hongkong11003     | 101 Switching Protocols (255ms)
India           | india1813         | 101 Switching Protocols (171ms)
Japan           | japan80           | 101 Switching Protocols (302ms)
Rotterdam       | rotterdam1572     | 101 Switching Protocols (58ms)
Russia          | N/A               | Discord API failed.
Singapore       | N/A               | Discord API failed.
South Africa    | N/A               | Discord API failed.
Syndey          | N/A               | Discord API failed.
US Central      | N/A               | Discord API failed.
US East         | N/A               | Discord API failed.
US South        | N/A               | Discord API failed.
US West         | N/A               | Discord API failed.
```

**Pull request merged**